File: shared/src/commonMain/kotlin/org/nihongo/mochi/domain/kana/KanaRepository.kt
package org.nihongo.mochi.domain.kana

import kotlinx.serialization.json.Json
import kotlinx.coroutines.runBlocking
import org.nihongo.mochi.ui.games.taquin.NumberData
import org.nihongo.mochi.ui.games.taquin.NumberEntry

class KanaRepository(private val resourceLoader: ResourceLoader) {
    
    private val json = Json { ignoreUnknownKeys = true }
    
    private var hiraganaCache: List<KanaEntry>? = null
    private var katakanaCache: List<KanaEntry>? = null

    fun getKanaEntries(type: KanaType): List<KanaEntry> {
        return runBlocking {
            getKanaEntriesSuspend(type)
        }
    }

    suspend fun getKanaEntriesSuspend(type: KanaType): List<KanaEntry> {
        if (type == KanaType.HIRAGANA && hiraganaCache != null) return hiraganaCache!!
        if (type == KanaType.KATAKANA && katakanaCache != null) return katakanaCache!!

        val fileName = when (type) {
            KanaType.HIRAGANA -> "kana/hiragana.json"
            KanaType.KATAKANA -> "kana/katakana.json"
        }
        
        val jsonString = resourceLoader.loadJson(fileName)
        val result = try {
            json.decodeFromString<KanaData>(jsonString).characters
        } catch (e: Exception) {
            println("Error parsing kana data: ${e.message}")
            emptyList()
        }
        
        if (type == KanaType.HIRAGANA) hiraganaCache = result
        if (type == KanaType.KATAKANA) katakanaCache = result
        
        return result
    }

    suspend fun getNumberEntries(): List<NumberEntry> {
        return try {
            val jsonString = resourceLoader.loadJson("numbers.json")
            json.decodeFromString<NumberData>(jsonString).numbers
        } catch (e: Exception) {
            println("Error parsing numbers data: ${e.message}")
            emptyList()
        }
    }
}

File: flutter_app/lib/repositories/kana_repository.dart
import 'dart:convert';
import '../models/kana.dart';
import '../services/resource_loader.dart';

class KanaRepository {
  final ResourceLoader _loader;
  List<KanaEntry>? _hiraganaCache;
  List<KanaEntry>? _katakanaCache;

  KanaRepository(this._loader);

  Future<List<KanaEntry>> getKanaEntries(KanaType type) async {
    if (type == KanaType.hiragana && _hiraganaCache != null) return _hiraganaCache!;
    if (type == KanaType.katakana && _katakanaCache != null) return _katakanaCache!;

    final assetPath = type == KanaType.hiragana
        ? 'assets/files/kana/hiragana.json'
        : 'assets/files/kana/katakana.json';
    final localPath = type == KanaType.hiragana ? 'kana/hiragana.json' : 'kana/katakana.json';

    try {
      final jsonString = await _loader.loadString(localPath, assetPath: assetPath);
      
      final Map<String, dynamic> jsonData = json.decode(jsonString);
      final result = KanaData.fromJson(jsonData).characters;

      if (type == KanaType.hiragana) _hiraganaCache = result;
      if (type == KanaType.katakana) _katakanaCache = result;

      return result;
    } catch (e) {
      print("Error loading kana data: $e");
      return [];
    }
  }

  Future<List<NumberEntry>> getNumberEntries() async {
    try {
      final jsonString = await _loader.loadString('suuji.json', assetPath: 'assets/files/suuji.json');
      final Map<String, dynamic> jsonData = json.decode(jsonString);
      return NumberData.fromJson(jsonData).numbers;
    } catch (e) {
      print("Error loading numbers: $e");
      return [];
    }
  }
}
